### DISCUSSION TITLE
GSoC Simplification project 2022

### DISCUSSION THREAD

**Discussion by martinfleis** [pysal/momepy#361] [General]
This year we have 2 GSoC contributors @gregmaya and @gsagostini working on the issue of street network simplification. This thread can be used for a general discussion during the summer.

**Running meeting notes** [here](https://hackmd.io/QKzsOzMORF6V4hbl5q386A).

**The original resources linked from the project idea page:**

Some of the existing tools offer partial funcionality:

- [`momepy.preprocessing` module](http://docs.momepy.org/en/stable/api.html#preprocessing)
- [`osmnx.simplification` module](https://osmnx.readthedocs.io/en/stable/osmnx.html#module-osmnx.simplification)
- [`cityseer`](https://cityseer.benchmarkurbanism.com/guide/#graph-cleaning) cleaning tools (note that `cityseer` has an incompatible license)

Additional resources:
- [@martinfleis's experiments](https://github.com/martinfleis/network_simplification)

**Comment by martinfleis** [comment]
@gregmaya @gsagostini Keep an eye on the work @dabreegster is doing over at https://github.com/a-b-street/osm2streets/issues/16 and https://github.com/a-b-street/osm2streets/pull/15. He will join the call next week to explain it a bit but there's some overlap already (you just need to be able to read Rust, which should be relatively okay).

**Reply by dabreegster** [reply]
Hi Greg and Gabriel! If you take a peek at these and have questions before the meeting, let me know. One way that effort might be immediately helpful is to collect a set of [test cases](https://github.com/a-b-street/osm2streets/tree/main/tests) of difficult street networks in OSM

**Reply by gregmaya** [reply]
Hi @dabreegster ! the A-B Street sounds like a very nice project! I happened to have worked on street design quite a bit a few years ago, where our main focus was to prioritize pedestrian experience rather than creating car-oriented/optimized junctions. Innevitebaly Aldo Van-Eyck was an early reference. see [this example](https://www.frontiersin.org/files/Articles/270518/fpsyg-08-01130-HTML/image_m/fpsyg-08-01130-g003.jpg). 
Anyway, I'll make sure I read through your efforts and hopefully, we can combine ideas for the best solution (not too far in the future!)

**Reply by dabreegster** [reply]
Listing out more links that may be helpful:

- https://github.com/a-b-street/osm2lanes turns tags from one OSM way into lanes left-to-right. It doesn't handle/understand parallel ways
- https://github.com/a-b-street/osm2streets/ (much more in-progress) generates geometry per road and intersection like https://github.com/a-b-street/osm2streets/blob/main/tests/src/borough_sausage_links/raw_map.json. The goal is for this library to have a bunch of transformations that collapse parallel ways
- https://dabreegster.github.io/talks/map_model_v2/slides.html motivates some of the simplifications

**Comment by csebastiao** [comment]
Here is an adaptation of the OSMnx simplification process to simplify a bicycle network that can be extracted through edges' attributes. We add the ability to discriminate OSM attributes to the simplification process of OSMnx:  https://github.com/anerv/NERDS_osmnx
Everything is based on OSMnx/OSM structure and is a network-based simplification to remove intersitial node here to preserve the geometry.

**Comment by jGaboardi** [comment]
One particularly ridiculous example (that also happens to be several miles from my home) is the Tom Moreland Interchange of I-85 and I-285 (a.k.a. [Atlanta Spaghetti Junction](https://www.georgiaencyclopedia.org/articles/counties-cities-neighborhoods/atlanta/spaghetti-junction_001/)).

Lon: -84.25986
Lat: 33.8920009

**Reply by jGaboardi** [reply]
Another weirdness up the street from where I used to live in central Atlanta is the duplication of Linden Ave NW. It is two parallel segments; and both are bi-directional. This is problem in the computational sense, both from routing & morphological perspectives, and actual usability (When first living in the area I took the wrong Linden Ave NW several times while coming home and had to detour up the one street to get back on track). From a morphological perspective, if we simplify with some weight given to attributes then considering the name would merge these parallel segments into one, which would then be ignoring the block in between (office building and small park).

Lon: -84.3885525
Lat: 33.7699201

<img width="1181" alt="Screen Shot 2022-06-17 at 10 56 45 AM" src="https://user-images.githubusercontent.com/8590583/174324167-108b5979-15a0-407e-88d2-be192a7e73b5.png">

<img width="1181" alt="Screen Shot 2022-06-17 at 11 01 21 AM" src="https://user-images.githubusercontent.com/8590583/174325538-d2078e3c-a792-494b-843f-d483b8dc980b.png">

**Comment by songololo** [comment]
Hi All,

Following this with interest as this is a pressing issue when using OSM sources for rigorous centrality methods (easily distorted by redundant topology or by conflation of geometry and topology.)

Re: cityseer's methods, I would be willing to discuss relaxing the license for the graphs module from AGPLv3 to maybe BSD3 or such if doing so would help find common ground amongst the packages?

For background, the issues I was trying to address via cityseers' methods:
- graph manipulation is easier to do in networkx than in dataframes, e.g. for London there are probably more than a million nodes affected (i.e. additions, deletions), so when cleaning and then decomposing the network then adding / removing rows from dataframes is quite slow whereas networkx is perfectly fine. networkx also makes it easier and more intuitive to explore neighbours and enact slightly more complex algorithmic situations.
- With cityseer there has been a lot of effort to cleanly separate the geometry from the topology. Geometry is abstracted to `shapely` `LineString` objects and these are then retained throughout the graph cleaning and manipulation process. i.e. the geometry is welded when stripping out redundant nodes, and cleanly split when decomposing. Robustness of node x, y coordinates is taken into account. This is emphasised heavily in cityseer so that the visual representation of the network remains agreeable and so that topological / geometrical distortions are minimised for the sake of retaining the integrity of the centrality methods. Shortest path centrality methods should work fine on automatically consolidated graphs though I would still caution against for simplest path centrality methods because even minor geometrical artefacts can heavily impact these algorithms. That said, cityseer works quite hard to reduce these sorts of artefacts.
- The space syntax community still likes to work with dual graphs, so cityseer includes a utility for dual graph conversions. The geometry is also split / welded accordingly. That said, cityseer's centrality methods can calculate angular 'simplest' path centrality methods on either primal or dual representations.
- I did a lot of experimentation a few years back on consolidating parallel roadways, the earlier methods I used worked fairly well, though in the end I've gone with the current approach which projects intersections to pierce adjacent parallel roadway segments and then simply rerun the consolidation methods to clean these up.
- There are so many ways to consolidate, and this is reflected in cityseer's configurable consolidation methods (crawling or not, min max nodes, etc.). On the surface this may seem intimidating to end-users, though these methods can simply be wrapped in simpler user facing functions for those that want a one-method-fits-all approach.

Anyway, will help or offer feedback if I can, but also don't want to meddle or interrupt any work you're doing.

**Reply by songololo** [reply]
PS - just a heads-up that cityseer's API has recently changed to enhance code quality and support use of `GeoDataFrame` for data state management where possible. The graph cleaning guide is mostly unchanged except that certain methods have been renamed to more correctly match convention, i.e. `nX_etc` has become `nx_etc`.

**Reply by gregmaya** [reply]
Gareth, many thanks for all this.
Next week, I´ll be studying in-depth some of the processes and methods you describe above. 
Originally, my thinking was that I would create a classification to split and narrow down the different methods needed for each one of the areas where corrections need to be made. In theory, this should not just improve accuracy but also the speed of the end solution(s) - been that 'geometry-based' or 'network-based'.
Anyway, great to have you on board (or at least with an eye) here!

**Comment by martinfleis** [comment]
A good way of organising your work with exploratory notebooks etc is to have a dedicated repository for that. You can check what Thomas did last year here https://github.com/tastatham/gsoc_dask_geopandas_2021.

**Comment by dabreegster** [comment]
Check out near page 18 of https://gisintransportation.com/media/zsnn3oou/2022_04_19-gis-t-workshop-aegist-slides.pdf

**Reply by jGaboardi** [reply]
@dabreegster Thanks for sharing this! I went through the full slide deck. That thing is *dense*...

**Comment by martinfleis** [comment]
@gregmaya have a look at https://centerline.readthedocs.io/en/latest/

**Reply by gregmaya** [reply]
Sure, will do! 
.. on the other hand I've managed to make some early functions to fix round abouts. The code needs some cleaning before I can share anything. But at least it's some momentum building !

**Reply by gregmaya** [reply]
As I said previously, I've been making some progress on the simplification of roundabouts.
The overall idea is to be able to classify the different 'error types' (a taks still pending) and to go from that to a solution that is case-specific... (like the one here on roundabouts). This approach is most probably slower than a combination of nodes based on a density per zone but hopefully more accurate in that it tackles the different nuances of different errors. 

In any case, it would be great to have your input on this first approach (https://github.com/gregmaya/gsoc2022_network_simpl/blob/main/exploratory_notebooks/round_about_functions.ipynb)
it's clearly very early stages and there's a lot to learn but I've tried to be quite explicit about what I'm doing to help readability...
you'll see at the end a strange case which I don't understand why when using the last function I'm getting an error 
`could not broadcast input array from shape (0,23) into shape (0,)`
but all works well when I take the actions within that and do them separately... 

(ps. Still in a remote town in the depths of France, so apologies if the testing hasn't involved many other cases... Any API call is too much for my network)

**Reply by amorfinv** [reply]
very cool @gregmaya ...currently taking a look.. one quick suggestion referring to the```rab_center_points()``` function. Another option could be to use the center of the circle rather than the dissolved polygon. I will let you know if I have more questions/suggestions.

**Reply by amorfinv** [reply]
also, I think all the warnings might be due to incompatible dependencies and some may be fixed in the future

1) for the ```TqdmWarning``` I did a ```mamba install ipywidgets``` see [here](https://stackoverflow.com/questions/67998191/importerror-iprogress-not-found-please-update-jupyter-and-ipywidgets-although).
2) The ```ShapelyDeprecationWarning``` will keep popping up between shapely 1.8 and 2.0. see [here](https://gis.stackexchange.com/questions/433401/shapely-2-0-warning-convert-the-coords-to-a-numpy-array-instead-related-to-ap) or shapely [docs](https://shapely.readthedocs.io/en/stable/migration.html#creating-numpy-arrays-of-geometry-objects). You can just use their code to ignore them during testing.

I usually do something like [this](https://gis.stackexchange.com/questions/420046/shapely-deprecation-warning-message-when-plotting-geopandas-geodataframe):

```python
import warnings
from shapely.errors import ShapelyDeprecationWarning
warnings.filterwarnings("ignore", category=ShapelyDeprecationWarning)
```

**Reply by amorfinv** [reply]
Also, not sure about the ```could not broadcast input array from shape (0,23) into shape (0,)``` I suspect it is the kernel getting confused when things get ran in different orders. I just ran everything with ```run all``` option from vscode and it seemed to work for me.

**Reply by gregmaya** [reply]
All makes total sense. Thanks @amorfinv 
I'll try to collect many other views (plus some of my own) stepping away from the task for a couple of days... But good to hear that you managed to run it and make sense of what was happening along the way!

**Reply by amorfinv** [reply]
I looked at the code a bit more and it is great. I have some comments that we can discuss at next week's meetings.

- I think ```roundabout_simpl()``` should only have the ```edges_gdf``` as the required argument. passing edges and polys seems a bit redundant since ```polys are``` created from edges.
- I am wondering if selecting only areas that are smaller than the roundabout in ```selecting_rabs_from_poly()``` should be the way forward...I can imagine that there is a case where the area of the roundabout is smaller than adjacent areas that you would like to collapse. However, maybe that would get fixed when you collapse parallel streets.
- I think there might be a simpler logic for```selecting_incoming_lines```. If I understood the process correctly, you extend each ```incoming``` ```LineString``` to the ```center_pt```. Then you check for duplicate geometry because several lines may have been cut at the same node. Finally, with ```COINS``` you decide which attributes to keep in the extended lines. My proposed logic would be:

1) find all unique points where the lines need to be extended to ```center_pt```. Should be relatively simple I think
2) create a ```gdf``` with ```LineString```s from cut points to ```center_pt```.
3) merge the ```gdf``` from **2** with ```incoming```.
4) do a coins magic to get the stroke_groups
5) copy attributes to new edges from an edge with the same stroke_ group. This can get a bit tricky because you need to get the attributes of the edge that shares the same stroke group and the same node that is not ```center_pt```.

**Reply by gregmaya** [reply]
Thanks for the feedback @amorfinv !
Definitely some points to discuss there.. I'll try to keep it short for some of your points to leave room for discussion collectively next week.
- I decided not to include the _polygonise_ into the first function thinking that those polygons could also serve some of the other upcoming functions. As we've been seeing with @martinfleis analysing their shape could also help us classify some of them. Now, if this doesn't end up to be a good approach it makes total sense to include the _poligonization_ into the function
- I think you're right about possible scenarios where adjacent polygons to the roundabouts are larges in size but you hit the nail on the head thinking that most likely those would become parallel and therefore tackled at a later stage. This is clearly start to suggest an important order of the function going forward.
- Finally, on your last point, it's worth mentioning that I only do the COINS analysis for the cases where more than one incoming line arrives in the same point to the polygon. The rest, which is actually the vast majority of the cases, the calculation if pretty straight forward. But happy to discuss all possible implications.

I admit I seemed to have fallen back into using some non-convetional ways for some processes because I was hitting errors. I.e. no sjoin predicate 'disjoin', or mapping attributes from one gdf to another without having to delete and re concatenate...etc... All things to improve. (Plus the very logical idea of having the center point of the actual circle you mentioned earlier)

**Reply by jGaboardi** [reply]
@gregmaya I took a look through your notebook. Pretty good progress so far and I agree with the points that @amorfinv raised. Next week I will be at SciPy 2022, so I won't be in our meeting, but I will of course see messages here and gitter.

**Reply by martinfleis** [reply]
@gregmaya which notebook do you want me to look at? The one linked above does not exist anymore.

**Reply by gregmaya** [reply]
@martinfleis I just pushed a commit to this network https://github.com/gregmaya/gsoc2022_network_simpl/blob/main/iv_finetuning_selecting_rabs.ipynb
(I see there's a warning for a 'long notebook' displayed because I added a `.explore() ` at the end. If you prefer I can just delete and push again so it renders online).
In any case, this notebook eas actually testing things that were then implemented in the functions which are here: https://github.com/gregmaya/gsoc2022_network_simpl/blob/main/rabs_simplify.py
I hope it all makes sense.

**Reply by martinfleis** [reply]
I think I will need a walkthrough next week to properly understand the logic behind it but it looks like a good progress!

https://github.com/gregmaya/gsoc2022_network_simpl/blob/2ae9d2376122f4f97d0ad9d889d48d4ac561a7f8/rabs_simplify.py#L24-L26 - Why do you filter out the largest polygons? Do they cause some trouble?

https://github.com/gregmaya/gsoc2022_network_simpl/blob/2ae9d2376122f4f97d0ad9d889d48d4ac561a7f8/rabs_simplify.py#L35-L44 - this looks like a bottleneck. Keep it loopy for now to get the functionality there but I think we will be able to vectorise it a bit. Also, when looping use `itertuples` instead of `iterrrows`. It is much faster.

https://github.com/gregmaya/gsoc2022_network_simpl/blob/2ae9d2376122f4f97d0ad9d889d48d4ac561a7f8/rabs_simplify.py#L71 - dissolve is expensive. You can cast the group to mutlipolygon instead, the centroid will be the same.

https://github.com/gregmaya/gsoc2022_network_simpl/blob/2ae9d2376122f4f97d0ad9d889d48d4ac561a7f8/rabs_simplify.py#L85-L88 - shapely 2.0/latest will not work for some time, we need a few changes in geopandas before. You can use this code instead for now https://github.com/pysal/momepy/blob/f6dd6973f8952ab6dba94f0377395c7cc9ec6f53/momepy/shape.py#L272-L451

There are some other bits that are worth looking into but I'd rather first understand what they're for :).

**Reply by gregmaya** [reply]
Awesome! Thanks, @martinfleis. I'll implement some of those suggested changes and would hopefully have some more progress next week. 
Overall the approach has been to make something 'work' and then (now) to go back and polish said function to accommodate for more nuances. 

Quick notes:
- We discussed last week how large roundabouts had two conditions : 

> 1. from a certain point they were urban blocks (many times squares) rather than transport results
> 2. given their size they used to also select adjacent polygons which shouldn't have 

- I agree with the vectorized version for the _hausdorff_distance_ . Given it was the last thing I added it's definitely there to test the outcome... might need to think about how to turn it into a more optimal way if we were to keep it. 
- I will make sure to add your other suggestions too before moving on too far and forgetting about them.

**Reply by martinfleis** [reply]
I would leave a threshold between what is a roundabout and what is a square as a parameter. It can be left to individual case studies to decide.

**Comment by gregmaya** [comment]
@gsagostini, I'm trying to create a list of cities/towns with enough variation in size and [degree of urbanization](https://ghsl.jrc.ec.europa.eu/ucdb2018visual.php#lnlt=@22.59373,10.54688,2z&v=201&ln=1&cls=0&HDC=na) as we discussed. I have been spending quite some time today trying to get such classification and it seems like the latest it's not quite fully public yet (see [this](https://ghsl.jrc.ec.europa.eu/download.php?ds=DUC) ). Do you/or anyone happens to have an older version by any chance?

**Reply by gsagostini** [reply]
@gregmaya we can probably use the Urban Centre Database for now ([this](https://ghsl.jrc.ec.europa.eu/download.php?ds=ucdb)). It does not have the actual urbanization class, but it has built-up area and population criteria, let alone continent tags, so we can use it for an initial sample. I have used [SMOD](https://ghsl.jrc.ec.europa.eu/download.php?ds=smod) in the past which has the urbanization class, but it is a raster-type classification by square mile across the world and not by urban center, which I think gets too complicated for us.

**Reply by martinfleis** [reply]
Guys, you are severely overthinking sample selection. Pick 5 random cities you like and make it work. Then we can think about fancy stuff like DoU. 

Edit: make it work at 1 large city and you are done, in fact.

**Comment by gsagostini** [comment]
July 12 meeting with @amorfinv @gregmaya and I:

- We have some historical network data to test our code on (keep private for now)
- Gabriel: working on an improved version of consolidate intersections from OSMnx. Adding support for undirected graphs. Considering using pandana to identify simultaneously all nodes that have small shortest network distance to another node, must check whether pandana can be included in momepy. Will post a notebook with examples and comparisons to the discussion.
- Greg: working on roundabout identification and consolidation. Notebooks and prior feedback in the thread above. Good graphs showing the choice of default parameters. We discussed how identifying roundabouts can also consider the area of the polygon (too big roundabouts are not necessarily to be simplified). Will check how useful it is to identify not only the roundabout but also small adjacent polygons to it---currently does it, may be worth to not do it and then call consolidate intersections on those polygons.

**Reply by martinfleis** [reply]
Thanks! I'll be back next week so will go through all your stuff in detail then. Quickly - pandana can become an optional dependency of momepy, that is no problem. It may not be the first time in the PySAL ecosystem (@knaaptime may know more) but is surely used within some of our workflows anyway.

**Reply by jGaboardi** [reply]
Yep, `pandana` is an [optional dependency](https://github.com/pysal/segregation/blob/b3b7ce1cc027132b752d6901f694ed1ad4dc01a9/segregation/network/network.py#L66-L74) in `segregation`.

**Reply by martinfleis** [reply]
@gsagostini is there anything we can look at before the next call?

**Reply by gsagostini** [reply]
@martinfleis I just pushed [this notebook](https://github.com/gsagostini/momepy/blob/main/gsocc/notebooks/22-06-30_intersections.ipynb) where I work through a function for intersection consolidation that is based solely on network distances. I compared it against the [OSMnx simplification module](https://github.com/gboeing/osmnx/blob/main/osmnx/simplification.py).

Key idea: 1) user gives tolerance to merge "close" nodes; 2) remove edges whose length is above this tolerance from the graph; 3) isolated nodes are the good ones i.e. those which do not have another node too close by so we can ignore them; 4) the remaining nodes form a disconnected graph, each component of that graph should be independently simplified into a single node; 5) rebuild edges. 

1. There is no need to treat nodes as a GeoDataFrame and do spatial operations of buffering and union. In the osmnx function, the nodes which are close in Euclidean distance are identified and afterwards a second network computation is made, so we avoid that redundancy.
2. The "tolerance" distance to merge nearby nodes is measured in network distance, not Euclidean distance. This can be easier to estimate.
3. Supports undirected graphs.
4. Faster for examples I have tried.
5. Because we start from disconnected clusters of nodes where we need to do simplification (see the first image in cell 3), we could classify those problems independently. Tying in the work of @gregmaya we could experiment with running the roundabout identification function not on the whole network, but only in polygons around these disjoint node clusters.

The edge reconstruction still needs work and hopefully I will have it better by the end of the week. I also used NetworkX for now because I couldn't directly use pandana to find graph components, but maybe that could also work (faster).

**Reply by martinfleis** [reply]
That is smart! And it should also directly support non-planar networks (tunnels etc) right? Hope you can figure out the edge reconstruction, this is promising!

**Reply by gregmaya** [reply]
Sounds like a proper classifier is ever more needed!!

**Reply by gsagostini** [reply]
@martinfleis yes it does not take any planarity into account---the network in the example has a bridge and it works just fine in that one. Classifying the issues would be nice because maybe some issues have better solutions than the standard consolidation (e.g. roundabouts). 

@gregmaya do you know where in your notebooks you did the edge reconstruction (i.e. connecting adjacent, unchanged nodes to the new roundabout node)? I would love to take a look and see I can find out what I am missing.

**Reply by martinfleis** [reply]
> where in your notebooks you did the edge reconstruction

When I was doing that last year, I did this
https://github.com/martinfleis/network_simplification/blob/208d5b985f86d483a8428802dfb98f759adbbb37/consolidate.py#L328-L343

**Reply by gregmaya** [reply]
In my case, I was starting the reconstruction with a set of gdfs which you might not have (the original geometry edges a preselected set of lines to extend and the polygons that originally make up for the roundbouts).
in any case have a look a this function :
https://github.com/gregmaya/gsoc2022_network_simpl/blob/dc8119d7bab1dd411164cc2551f85b4c8fe7e047/rabs_simplify.py#L159-L175

**Reply by gsagostini** [reply]
I [updated the notebook](https://github.com/gsagostini/momepy/blob/main/gsocc/notebooks/22-06-30_intersections.ipynb) with a fully-finished function which also reconstructs the edges. It's a method that works for undirected graphs and is agnostic to planarity. I think there are different methods to reconstruct edges (differing on how expensive they are, and for what purpose they will be valuable), so I included two methods for now:

1. An "as-the-crow-flies" method that ignores any pre-existing geometry indication. This is the fastest reconnection method and, most importantly, it preserves adjacency. For folks studying network properties (degree distribution, random walks, motif analysis) that do not rely in the edge geometry, this is ideal.
2. A "linear" method that extends the edge geometry from the original endpoint to the cluster centroid corresponding to that endpoint. This approximates morphology a bit better.

I will work on including something similar to your code @martinfleis on the spider-like reconstruction because it does look better than the linear way. And since I imagine polygons will be running around at some point I will also think of how to use them to do an "averaging two edges between clusters" to get a single-line simplification of, say, a two-lane street.

We should discuss more on Tuesday, but I envisioned this function as a first simplification step. For some folks, it may be enough. But if additional precision is required, we can as I said above classify the imperfections and apply specific methods to simplify e.g. bridges, interchanges, etc.

**Reply by amorfinv** [reply]
hey @gsagostini I did not get a chance to dig deep into your code last week. I have some time tomorrow and I was wondering if I should look into this [notebook](https://github.com/gsagostini/momepy/blob/main/gsocc/notebooks/22-06-30_intersections.ipynb) or if there is a more updated one.

-Andres

**Reply by gsagostini** [reply]
That's my last polished work, would love some feedback on it. I am cleaning up a spider-web simplification to avoid overlapping lines, but may only push it by tomorrow

**Reply by amorfinv** [reply]
Hey @gsagostini I finally got a chance to review your [notebook](https://github.com/gsagostini/momepy/blob/main/gsocc/notebooks/22-06-30_intersections.ipynb). It is some nice work. I really liked your coding style and may steal a couple of tricks from you in the future. I think it is ready for a draft PR. Here are my comments:

1. At the moment ```consolidate_intersections_nx()``` would only work with ```osmnx``` graphs since it depends on their naming conventions. For example, in ```get_rebuilt_edges()``` the renaming of the columns requires that a "from" and "to" column get defined. I would recommend to find a way to make it source agnostic. Or maybe you can make it a ```momepy``` built  ```gdf```. With ```momepy.nx_to_gdf()``` (see [here](http://docs.momepy.org/en/stable/generated/momepy.nx_to_gdf.html#momepy.nx_to_gdf)) The "too" and "from" columns from ```osmnx``` get replaced by "node_start" and  "node_end".  I think the ```consolidate_intersections_nx()```  also depends on having the "u" , "v", "key" format for indexing. 
2. Related to (1), perhaps it makes sense that ```momepy``` performs some initial formatting so that you don't depend on an outside naming convention. Or maybe this pre-formatting happens outside the function? I am not sure what the right answer is yet but it is something to think about.
3. Another case where the function depends on ```consolidate_intersections_nx()``` depends on ```osmnx``` is on the ```if``` statement at the end where you convert the ```nx``` graph to an edge ```gdf```. I think you should use ```momepy``` here.
4. In the case of a self loop in the  ```linear_simplification()``` function there would be an error because ```line_coords``` is not defined. It did not seem to happen in the current networks you are testing so the error never appeared.
5. Related to the self loop for a cluster. I had a hard time visualizing when that would happen. So perhaps you can come up with an example. It would be a good idea to have this in the tests.
6. I think the use of ```deepcopy``` is generally undesired as it can be quite expensive especially for larger graphs. However, if a deepcopy is needed then it is needed. I am unsure if there is a way to avoid but it can be something to think about.

Those are my main comments. I also had an unrelated comments that you can ignore:
7. What does the ```fill_edge_geometry``` argument in ```osmnx.graph_to_gdfs()```  mean. I briefly looked at the ```osmnx``` documentation but it did not really make sense. Maybe once you switch to ```momepy``` functions then it won't matter.

Looking forward to the spider simplification! Let me know if you want to discuss something before the meeting next week.

-Andres

**Reply by gsagostini** [reply]
Hi @amorfinv thanks a lot for the feedback.

1-3. Will definitely make the node and edge keywords agnostic. I am thinking, for starting, to just allow those as input keywords. I will check the moment nx to gdf and use it instead of the osmnx, as they should be sort of interchangeable.

4-5. Good catch! I did had a lot of trouble coming up with that case, since the line would form a closed loop. I will try my best to find a place where this happens and include it on a test.

6. Noted, will do my best to avoid. Issue for me is I wanted to copy a graph in which I could remove edges without changing the original graph, and that was behaving wacky with a shallow copy. Perhaps I can try storing the original graph's nodes, edges, and data separately and then recreating it rather than keeping that object immutable. Will give it a try, and compare runtimes.

7. Fill edge geometry is quite important when the user passes an osmnx graph.  Some edges do not have as attributes their geometry---essentially, when a user calls osmnx.plot_graph the function just makes straight lines when the geometry is missing. I imagine this should be done outside the function even, as it is kind of a data pre-processing, but many folks using osmnx graphs may need it.

I should have the spider simplification too by tomorrow and then, along with addressing these points, I will draft a PR.

**Reply by amorfinv** [reply]
Perhaps if you don't find a self-loop example you can create one.

Ah, I had no clue that I could just ignore edge geometry with ```osmnx```. 

Looking forward to PR

**Reply by martinfleis** [reply]
Just a side note. If you need momepy to behave slightly differently when converting to graph and back, we can change the existing implementation in some compatible way.

**Reply by gsagostini** [reply]
I created a [draft PR](https://github.com/pysal/momepy/pull/377) with the finalized version of this function which we can look at during the meeting today. The comments above were mostly addressed and the spider-web simplification is finished. I tried to document my assumptions on network attributes, and I think maybe we can drop some of them by preprocessing.

**Comment by gregmaya** [comment]
I just uploaded [this nothebook](https://github.com/gregmaya/gsoc2022_network_simpl/blob/main/27-07-22_rabs_simpl_output.ipynb) where I displayed the outputs of the _rabs_simplify_ up to this point. Unfortunately, rendering the folium maps is too much to be rendered on GitHub but it's definitely useful to understand the output in context.
As you can see there are still quite a few warning outputs when running the function, which becomes very obvious when testing performance/time. 
- Overall, 689 roundabouts were modified and the avg. time is just over 9 secs.

**Reply by gregmaya** [reply]
Guys, apologies for not having been able to create a pull request before the end of today. The functions are ready and the .RST file too but I had some issues at home today that I needed to take care before the end of the week which delayed my work. I jop I'll be doing said pull request this weekend...
Only tests are missing.

**Reply by gregmaya** [reply]
Here is a [DRAFT PULL REQUEST](https://github.com/pysal/momepy/pull/371). The 'draft' sounded like a safe way to go about this first attempt.
Happy to hear your comments!

**Comment by gregmaya** [comment]
@amorfinv as promised I recreated what we were discussing yesterday in regards to the COINS outputs. 
Currently, the `stroke_gdf` outputs the group split that we need to assign the new extension to the correct incoming LineString.
@martinfleis had suggested that in order to simplify the processes we would use `stroke_attribute()`. 

For reference, I left in RAW cells the functions that are affected in the original `roundabout_simplification()` function.
Here's [the notebook](https://github.com/gregmaya/gsoc2022_network_simpl/blob/main/enh_examples.ipynb)

If you can have a look and maybe suggest what could be done better in the use of COINS it would be great.
Many thanks!

**Reply by martinfleis** [reply]
Can you share `"/Users/gregoriomaya/Desktop/GSoc_2022/gsoc2022_network_simpl/data/madrid2062.gpkg"`?

**Reply by gregmaya** [reply]
I believe you can download it here https://github.com/gregmaya/gsoc2022_network_simpl/blob/main/data/madrid2062.gpkg

**Reply by amorfinv** [reply]
Thanks!  Will take a look later today or tomorrow

**Reply by amorfinv** [reply]
Hey @gregmaya .. there is something definitely odd. In the last bit of your of notebook there is a for loop that cycles through two gdfs. In the first one coins works as expected but there is something odd with the second.. I think I'll have to dig into it a bit more tomorrow morning..

**Reply by amorfinv** [reply]
Hey @gregmaya I now see why it is giving strange results and it is because of the ```pd.concat()``` that you call in

```python 
gs = gpd.GeoSeries(pd.concat([x.geometry, x.line]), crs=incoming_many.crs)
```

In the second loop of your example it stacks things with the same indices so then COINS doesn't know what to do for the attributes because it depends on unique indices to match the series back to the original ```gdf```. (@martinfleis I guess we could make it work to not care about the indices but unsure if it is worth it). The solution can be to just modify the line to be

```python 
gs = gpd.GeoSeries(pd.concat([x.geometry, x.line], ignore_index=True), crs=incoming_many.crs)
```

or if you care about the indices you can save them as a column and reset with

```python
gdf.reset_index(inplace=True)
```

**Reply by gregmaya** [reply]
Aha !! 
Interesting... Thanks for looking into it! @amorfinv 
I'll see if I can figure out a way to make things simpler knowing this now !

**Reply by martinfleis** [reply]
@amorfinv nice catch! We should do at least something about it. This is a bug on our side. Ideally make it index-agnostic but the least what we should do is to check for duplicates in the index and raise an error.

**Reply by amorfinv** [reply]
I thought about it a bit more and I think making it index agnostic should be relatively easy so I will try to make the change at some point this weekend.

**Reply by amorfinv** [reply]
@gregmaya done! In the end it is just a one line change haha..your original code now works on one of my [branches](https://github.com/amorfinv/momepy/tree/patch-coins).

**Reply by gregmaya** [reply]
Well that was fast! 
Awesome!! ... Now the never ending improvement of my PR !!

**Reply by martinfleis** [reply]
I will merge @amorfinv's PR so you can build on top of that, having the issue resolved. Once it is merged, you will need to do:

```
# ensure you are on your branch
git fetch upstream
git merge upstream/main
```

**Comment by gregmaya** [comment]
In regards to the `center_type = "mean"` in `_rabs_center_points` 
- I created this [notebook](https://github.com/gregmaya/gsoc2022_network_simpl/blob/main/enh_examples-means.ipynb) where, to my surprise, the `"mean"` Point actually works for cases with and without adjacent polygons.
- I, therefore, included a **check** on the current PR that asserts its use.

**Comment by martinfleis** [comment]
Hi @gsagostini, how is your part going right now? Can you give us a short update so we know what to expect and where to help given the submission deadline is in 5 days? I am also happy to find a time for another call this week if you think that may be useful.

**Reply by gsagostini** [reply]
Hi! My bad for being a bit absent from this, wrapping up the notebook with examples and the tests for the PR. Let's schedule a call this week---any time Friday?---, I think this would be helpful.

**Reply by martinfleis** [reply]
What would be the earliest you can do on Friday given the time zone difference? I need to leave the office sooner than normally and have another call at our _usual_ time.

**Reply by gsagostini** [reply]
is 8am EST (30min) too late already over there?

**Reply by martinfleis** [reply]
That should be okay, if we stick to 30 mins. Got another call at 8:30am EST. After that it may be too late for me but maybe @jGaboardi or @amorfinv would be available.

I have created an even in the PySAL calendar:

GSoC Simplification (Gabriel)
Friday, 9 September · 14:00 – 14:30 (Prague)
Google Meet joining info
Video call link: https://meet.google.com/xxk-pyuf-gmp
Or dial: ‪(CZ) +420 234 610 631‬ PIN: ‪540 424 779‬#
More phone numbers: https://tel.meet/xxk-pyuf-gmp?pin=3245189651708

**Reply by jGaboardi** [reply]
I can join from 8:00-8:30 EST, but also have commitments after 8:30.

**Reply by amorfinv** [reply]
I am not sure I will be able to then..I am heading to the airport at that time..if I get past security on time I will join

**Comment by gregmaya** [comment]
@martinfleis @jGaboardi @amorfinv I just wanted to raise your attention to some flowcharts that I did to document the processes and visualize some of the mazes that we've been navigating (when thinking mostly at a geometry-based solution)... As I said there's some text missing there to explain a bit further but would appreciate your input, too. 

See [this link](https://github.com/gregmaya/gsoc2022_network_simpl/blob/main/README.md), which will be the one that will form the submission for the closure of GSOC2022 (when complete!)

**Reply by amorfinv** [reply]
congrats on the merge

**Reply by gregmaya** [reply]
OK so as I'm just making a little notebook for exemplifying the function, more ideas come for enhancements of that same function... 
I'm going to leave the idea here before it escapes my head as I have quite a few things to wrap up before the end of GSOC.

So, on the selection of adjacent polygons we're mostly basing that on three factors: 
1. for a polygon to be touching the actual roundabout AND
2. for it to be smaller (in area) than the roundabout it touches AND 
3. that it doesn't stretch further than a `diameter_factor` of the diameter of the roundabout.
This approach works a good amount of the time but also I've seen cases where points 2 and 3 are not met when in reality they should.

My hypothesis:
what if we use the number of forming edges* (i.e. number of original edges that formed the polygons) to determine if those touching polygons are 'triangle-like' (formed by 3 edges)?

*This extra parameter is being used in some of the other functions I've been working on, so it shouldn't be too much work to combine the ideas.

Hoping this makes sense and it does not just become a _note to myself_
